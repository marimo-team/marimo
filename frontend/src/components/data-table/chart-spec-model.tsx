/* Copyright 2024 Marimo. All rights reserved. */
import type { TopLevelFacetedUnitSpec } from "@/plugins/impl/data-explorer/queries/types";
import { mint, orange, slate } from "@radix-ui/colors";
import type { ColumnHeaderSummary, FieldTypes } from "./types";
import { asURL } from "@/utils/url";

const MAX_BAR_HEIGHT = 24; // px
const MAX_BAR_WIDTH = 28; // px
const CONTAINER_WIDTH = 120; // px
const PAD = 1; // px
const VARIABLE_WIDTH = `min(${MAX_BAR_WIDTH}, ${CONTAINER_WIDTH} / length(data('source_0')) - ${PAD})`;

const scale = {
  align: 0,
  paddingInner: 0,
  paddingOuter: {
    expr: "length(data('source_0')) == 2 ? 1 : length(data('source_0')) == 3 ? 0.5 : length(data('source_0')) == 4 ? 0 : 0",
  },
};

export class ColumnChartSpecModel<T> {
  private columnSummaries = new Map<string | number, ColumnHeaderSummary>();

  public static readonly EMPTY = new ColumnChartSpecModel([], {}, [], {
    includeCharts: false,
  });

  constructor(
    private readonly data: T[] | string,
    private readonly fieldTypes: FieldTypes,
    readonly summaries: ColumnHeaderSummary[],
    private readonly opts: {
      includeCharts: boolean;
    },
  ) {
    this.columnSummaries = new Map(summaries.map((s) => [s.column, s]));
  }

  public getHeaderSummary(column: string) {
    return {
      summary: this.columnSummaries.get(column),
      type: this.fieldTypes[column],
      spec: this.opts.includeCharts ? this.getVegaSpec(column) : undefined,
    };
  }

  private getVegaSpec<T>(column: string): TopLevelFacetedUnitSpec | null {
    if (!this.data) {
      return null;
    }
    if (typeof this.data !== "string") {
      return null;
    }

    const base: Omit<TopLevelFacetedUnitSpec, "mark"> = {
      data: {
        url: asURL(this.data).href,
      } as TopLevelFacetedUnitSpec["data"],
      background: "transparent",
      config: {
        view: {
          stroke: "transparent",
        },
        axis: {
          domain: false,
        },
      },
      height: 100,
    };
    const type = this.fieldTypes[column];

    switch (type) {
      case "date":
        return {
          ...base,
          mark: {
            type: "bar",
            color: mint.mint11,
            width: { expr: VARIABLE_WIDTH },
          },
          encoding: {
            x: {
              field: column,
              type: "temporal",
              axis: null,
              bin: true,
              scale: scale,
            },
            y: { aggregate: "count", type: "quantitative", axis: null },
            tooltip: [
              {
                field: column,
                type: "temporal",
                format: "%Y-%m-%d",
                bin: true,
                title: column,
              },
              {
                aggregate: "count",
                type: "quantitative",
                title: "Count",
                format: ",d",
              },
            ],
            // Color nulls
            color: {
              condition: {
                test: `datum["bin_maxbins_10_${column}_range"] === "null"`,
                value: orange.orange11,
              },
              value: mint.mint11,
            },
          },
        };
      case "integer":
      case "number": {
        const format = type === "integer" ? ",d" : ".2f";
        return {
          ...base,
          mark: {
            type: "bar",
            color: mint.mint11,
            size: { expr: VARIABLE_WIDTH },
            align: "right",
          },
          encoding: {
            x: {
              field: column,
              type: "nominal",
              axis: null,
              bin: true,
              scale: scale,
            },
            y: {
              aggregate: "count",
              type: "quantitative",
              axis: null,
              scale: { type: "linear" },
            },
            tooltip: [
              {
                field: column,
                type: "nominal",
                format: format,
                bin: true,
                title: column,
              },
              {
                aggregate: "count",
                type: "quantitative",
                title: "Count",
                format: ",d",
              },
            ],
            // Color nulls
            color: {
              condition: {
                test: `datum["bin_maxbins_10_${column}_range"] === "null"`,
                value: orange.orange11,
              },
              value: mint.mint11,
            },
          },
        };
      }
      case "boolean":
        return {
          ...base,
          mark: { type: "bar", color: mint.mint11 },
          encoding: {
            y: {
              field: column,
              type: "nominal",
              axis: {
                labelExpr:
                  "datum.label === 'true' || datum.label === 'True'  ? 'True' : 'False'",
                tickWidth: 0,
                title: null,
                labelColor: slate.slate9,
              },
            },
            x: {
              aggregate: "count",
              type: "quantitative",
              axis: null,
              scale: { type: "linear" },
            },
            tooltip: [
              { field: column, type: "nominal", title: "Value" },
              {
                aggregate: "count",
                type: "quantitative",
                title: "Count",
                format: ",d",
              },
            ],
          },
          layer: [
            {
              mark: {
                type: "bar",
                color: mint.mint11,
                height: MAX_BAR_HEIGHT,
              },
            },
            {
              mark: {
                type: "text",
                align: "left",
                baseline: "middle",
                dx: 3,
                color: slate.slate9,
              },
              encoding: {
                text: {
                  aggregate: "count",
                  type: "quantitative",
                },
              },
            },
          ],
        } as TopLevelFacetedUnitSpec; // "layer" not in TopLevelFacetedUnitSpec
      case "unknown":
      case "string":
        return null;
      default:
        return null;
    }
  }
}
