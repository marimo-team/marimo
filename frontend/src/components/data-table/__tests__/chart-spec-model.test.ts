/* Copyright 2024 Marimo. All rights reserved. */
import { describe, it, expect } from "vitest";
import { ColumnChartSpecModel } from "../chart-spec-model";
import type { ColumnHeaderSummary, FieldTypes } from "../types";

describe("ColumnChartSpecModel", () => {
  const mockData = "http://example.com/data.json";
  const mockFieldTypes: FieldTypes = {
    date: "date",
    number: "number",
    integer: "integer",
    boolean: "boolean",
    string: "string",
  };
  const mockSummaries: ColumnHeaderSummary[] = [
    { column: "date", min: "2023-01-01", max: "2023-12-31" },
    { column: "number", min: 0, max: 100 },
    { column: "integer", min: 1, max: 10 },
    { column: "boolean", true: 5, false: 5 },
    { column: "string", unique: 20 },
  ];

  it("should create an instance", () => {
    const model = new ColumnChartSpecModel(
      mockData,
      mockFieldTypes,
      mockSummaries,
      { includeCharts: true },
    );
    expect(model).toBeInstanceOf(ColumnChartSpecModel);
  });

  it("should return EMPTY for static EMPTY property", () => {
    expect(ColumnChartSpecModel.EMPTY).toBeInstanceOf(ColumnChartSpecModel);
    expect(ColumnChartSpecModel.EMPTY.summaries).toEqual([]);
  });

  it("should return header summary with spec when includeCharts is true", () => {
    const model = new ColumnChartSpecModel(
      mockData,
      mockFieldTypes,
      mockSummaries,
      { includeCharts: true },
    );
    const dateSummary = model.getHeaderSummary("date");
    expect(dateSummary.summary).toEqual(mockSummaries[0]);
    expect(dateSummary.type).toBe("date");
    expect(dateSummary.spec).toBeDefined();
  });

  it("should return header summary without spec when includeCharts is false", () => {
    const model = new ColumnChartSpecModel(
      mockData,
      mockFieldTypes,
      mockSummaries,
      { includeCharts: false },
    );
    const numberSummary = model.getHeaderSummary("number");
    expect(numberSummary.summary).toEqual(mockSummaries[1]);
    expect(numberSummary.type).toBe("number");
    expect(numberSummary.spec).toBeUndefined();
  });

  it("should return null spec for string and unknown types", () => {
    const model = new ColumnChartSpecModel(
      mockData,
      mockFieldTypes,
      mockSummaries,
      { includeCharts: true },
    );
    const stringSummary = model.getHeaderSummary("string");
    expect(stringSummary.spec).toBeNull();
  });

  it("should handle special characters in column names", () => {
    const specialFieldTypes: FieldTypes = {
      "column.with[special:chars]": "number",
    };
    const specialSummaries: ColumnHeaderSummary[] = [
      { column: "column.with[special:chars]", min: 0, max: 100 },
    ];
    const model = new ColumnChartSpecModel(
      mockData,
      specialFieldTypes,
      specialSummaries,
      { includeCharts: true },
    );
    const summary = model.getHeaderSummary("column.with[special:chars]");
    expect(summary.spec).toBeDefined();
    expect((summary.spec?.encoding?.x as { field: string })?.field).toBe(
      "column\\.with\\[special\\:chars\\]",
    );
  });

  describe("snapshot", () => {
    const fieldTypes: FieldTypes = {
      ...mockFieldTypes,
      a: "number",
    };

    it("url data", () => {
      const model = new ColumnChartSpecModel(
        mockData,
        fieldTypes,
        mockSummaries,
        { includeCharts: true },
      );
      expect(model.getHeaderSummary("date").spec).toMatchSnapshot();
    });

    it("csv data", () => {
      const model = new ColumnChartSpecModel(
        `data:text/csv;base64,${btoa("a,b,c\n1,2,3\n4,5,6")}`,
        fieldTypes,
        mockSummaries,
        { includeCharts: true },
      );
      expect(model.getHeaderSummary("a").spec).toMatchSnapshot();
    });

    it("csv string", () => {
      const model = new ColumnChartSpecModel(
        "a,b,c\n1,2,3\n4,5,6",
        fieldTypes,
        mockSummaries,
        { includeCharts: true },
      );
      expect(model.getHeaderSummary("a").spec).toMatchSnapshot();
    });

    it("array", () => {
      const model = new ColumnChartSpecModel(
        ["a", "b", "c"],
        fieldTypes,
        mockSummaries,
        { includeCharts: true },
      );
      expect(model.getHeaderSummary("a").spec).toMatchSnapshot();
    });
  });
});
