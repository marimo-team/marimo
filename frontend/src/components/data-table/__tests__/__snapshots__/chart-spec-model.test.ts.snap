// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ColumnChartSpecModel > snapshot > array 1`] = `
{
  "background": "transparent",
  "config": {
    "axis": {
      "domain": false,
    },
    "view": {
      "stroke": "transparent",
    },
  },
  "data": {
    "values": [
      "a",
      "b",
      "c",
    ],
  },
  "encoding": {
    "color": {
      "condition": {
        "test": "datum["bin_maxbins_10_a_range"] === "null"",
        "value": "#cc4e00",
      },
      "value": "#027864",
    },
    "tooltip": [
      {
        "bin": true,
        "field": "a",
        "format": ".2f",
        "title": "a",
        "type": "nominal",
      },
      {
        "aggregate": "count",
        "format": ",d",
        "title": "Count",
        "type": "quantitative",
      },
    ],
    "x": {
      "axis": null,
      "bin": true,
      "field": "a",
      "scale": {
        "align": 0,
        "paddingInner": 0,
        "paddingOuter": {
          "expr": "length(data('source_0')) == 2 ? 1 : length(data('source_0')) == 3 ? 0.5 : length(data('source_0')) == 4 ? 0 : 0",
        },
      },
      "type": "nominal",
    },
    "y": {
      "aggregate": "count",
      "axis": null,
      "scale": {
        "type": "linear",
      },
      "type": "quantitative",
    },
  },
  "height": 100,
  "mark": {
    "align": "right",
    "color": "#027864",
    "size": {
      "expr": "min(28, 120 / length(data('source_0')) - 1)",
    },
    "type": "bar",
  },
}
`;

exports[`ColumnChartSpecModel > snapshot > csv data 1`] = `
{
  "background": "transparent",
  "config": {
    "axis": {
      "domain": false,
    },
    "view": {
      "stroke": "transparent",
    },
  },
  "data": {
    "values": [
      {
        "a": 1,
        "b": 2,
        "c": 3,
      },
      {
        "a": 4,
        "b": 5,
        "c": 6,
      },
    ],
  },
  "encoding": {
    "color": {
      "condition": {
        "test": "datum["bin_maxbins_10_a_range"] === "null"",
        "value": "#cc4e00",
      },
      "value": "#027864",
    },
    "tooltip": [
      {
        "bin": true,
        "field": "a",
        "format": ".2f",
        "title": "a",
        "type": "nominal",
      },
      {
        "aggregate": "count",
        "format": ",d",
        "title": "Count",
        "type": "quantitative",
      },
    ],
    "x": {
      "axis": null,
      "bin": true,
      "field": "a",
      "scale": {
        "align": 0,
        "paddingInner": 0,
        "paddingOuter": {
          "expr": "length(data('data_0')) == 2 ? 1 : length(data('data_0')) == 3 ? 0.5 : length(data('data_0')) == 4 ? 0 : 0",
        },
      },
      "type": "nominal",
    },
    "y": {
      "aggregate": "count",
      "axis": null,
      "scale": {
        "type": "linear",
      },
      "type": "quantitative",
    },
  },
  "height": 100,
  "mark": {
    "align": "right",
    "color": "#027864",
    "size": {
      "expr": "min(28, 120 / length(data('data_0')) - 1)",
    },
    "type": "bar",
  },
}
`;

exports[`ColumnChartSpecModel > snapshot > csv string 1`] = `
{
  "background": "transparent",
  "config": {
    "axis": {
      "domain": false,
    },
    "view": {
      "stroke": "transparent",
    },
  },
  "data": {
    "values": [
      {
        "a": 1,
        "b": 2,
        "c": 3,
      },
      {
        "a": 4,
        "b": 5,
        "c": 6,
      },
    ],
  },
  "encoding": {
    "color": {
      "condition": {
        "test": "datum["bin_maxbins_10_a_range"] === "null"",
        "value": "#cc4e00",
      },
      "value": "#027864",
    },
    "tooltip": [
      {
        "bin": true,
        "field": "a",
        "format": ".2f",
        "title": "a",
        "type": "nominal",
      },
      {
        "aggregate": "count",
        "format": ",d",
        "title": "Count",
        "type": "quantitative",
      },
    ],
    "x": {
      "axis": null,
      "bin": true,
      "field": "a",
      "scale": {
        "align": 0,
        "paddingInner": 0,
        "paddingOuter": {
          "expr": "length(data('data_0')) == 2 ? 1 : length(data('data_0')) == 3 ? 0.5 : length(data('data_0')) == 4 ? 0 : 0",
        },
      },
      "type": "nominal",
    },
    "y": {
      "aggregate": "count",
      "axis": null,
      "scale": {
        "type": "linear",
      },
      "type": "quantitative",
    },
  },
  "height": 100,
  "mark": {
    "align": "right",
    "color": "#027864",
    "size": {
      "expr": "min(28, 120 / length(data('data_0')) - 1)",
    },
    "type": "bar",
  },
}
`;

exports[`ColumnChartSpecModel > snapshot > url data 1`] = `
{
  "background": "transparent",
  "config": {
    "axis": {
      "domain": false,
    },
    "view": {
      "stroke": "transparent",
    },
  },
  "data": {
    "values": [],
  },
  "encoding": {
    "color": {
      "condition": {
        "test": "datum["bin_maxbins_10_date_range"] === "null"",
        "value": "#cc4e00",
      },
      "value": "#027864",
    },
    "tooltip": [
      {
        "bin": true,
        "field": "date",
        "format": "%Y-%m-%d",
        "title": "date",
        "type": "temporal",
      },
      {
        "aggregate": "count",
        "format": ",d",
        "title": "Count",
        "type": "quantitative",
      },
    ],
    "x": {
      "axis": null,
      "bin": true,
      "field": "date",
      "scale": {
        "align": 0,
        "paddingInner": 0,
        "paddingOuter": {
          "expr": "length(data('data_0')) == 2 ? 1 : length(data('data_0')) == 3 ? 0.5 : length(data('data_0')) == 4 ? 0 : 0",
        },
      },
      "type": "temporal",
    },
    "y": {
      "aggregate": "count",
      "axis": null,
      "type": "quantitative",
    },
  },
  "height": 100,
  "mark": {
    "color": "#027864",
    "type": "bar",
    "width": {
      "expr": "min(28, 120 / length(data('data_0')) - 1)",
    },
  },
}
`;
